### ISSUE TITLE
Add minItems property to Repeater

### ISSUE THREAD

**Issue by ayumihamsaki** [octobercms/october#3668]
Been searching for the solution to this issue and have decided to create an issue.

In October CMS when you create a Repeater field it will display like this:

![1](https://user-images.githubusercontent.com/17784082/43665710-034569a4-9769-11e8-97e9-ae4a0aece33d.png)

However this is very bad Web Design and many case studies have said not to create hidden fields which reduce user interactions.

I am wanting to create a repeater field that will show the user the first array without the need to click the add button, see below:

![2](https://user-images.githubusercontent.com/17784082/43665795-44254f34-9769-11e8-90f1-213b5c52db29.png)

Is this possible in October CMS or not and would it be a good idea to have this option if its not already there as standard.

Thanks in advance.

**Comment by LukeTowers** [comment]
@ayumihamsaki I don't know the reasoning behind this decision for sure, but I'm guessing that it doesn't display the first item by default right now because when you go to save it when you haven't actually entered anything into the field it would save an empty entry. This could cause problems, especially when you factor in fields having default values which would prevent a solution where the first item isn't saved if it's the only item and empty.

**Comment by ayumihamsaki** [comment]
@LukeTowers Hi I understand what you are saying so I thought I test out your comment, which produces the following database results:

### Hidden Menu

![1](https://user-images.githubusercontent.com/17784082/43677789-02ba089a-9800-11e8-9dca-ebec31997a98.png)

### Displayed Menu

![2](https://user-images.githubusercontent.com/17784082/43677797-112d77ea-9800-11e8-98de-3a55e7396021.png)

Both options are showing the JSON properties as having no results. I think some small coding could also check to see if the properties have values or not and if not then save the result in the database as `[]`.

I was kind of hoping to have a solution where we could have a choice in the Models fields/columns file to show the first array or not, like this for example:

```
extra_information:
    type: repeater
    show: true
    form:
        fields:
            added_at:
                label: Date added
                type: datepicker
            details:
                label: Details
                type: textarea
```

**Note:** The field `show` where we could set it to `true` or `false`.

Then giving developers a choice.

**Comment by ayumihamsaki** [comment]
@LukeTowers That idea sounds prefect.

**Comment by ayumihamsaki** [comment]
@LukeTowers 

Your idea sounds prefect with the `minItems` - would it get around the `default` issue?

Would it be a good idea to pre-display the repeater field with a `minItems: 0` and have an event listener that checks to see if the user has updated the first array field(s) or not before saving?

If the user has not changed the fields in the repeater field then save the repeater with: `minItems: 0` (same as hidding the repeater field at the moment).

If the user has changed the first array fields then set the `minItems: 1` and save the following data.

The `minItems` I take it can get can Increment and Decrement depending on the repeater rows.

I'm not sure what Validation checks will be needed?

**Comment by LukeTowers** [comment]
@ayumihamsaki the idea would be that the developer would set `minItems: 1` to display the first repeater item and it would be up to them to perform any validation that they might require to prevent that from being saved when it shouldn't.

**Comment by ayumihamsaki** [comment]
@LukeTowers 

I understand what you are saying. 

I would be incline to add an extra thing on top though. If the array fields are all blank values then save it to `[]`. If the developer has a `default` then yes they can create the extra validation. So the following examples:

```
extra_information:
    type: repeater
    show: true
    form:
        fields:
            added_at:
                label: Date added
                type: datepicker
            details:
                label: Details
                type: textarea
```

On save gives:

`[{"added_at":"","details":""}]`

Because **_ALL_** blank values can be saved instead as `[]` in database. **_Then most people would not need to change any code I think_**.

Then the default:

```
extra_information:
    type: repeater
    show: true
    form:
        fields:
            added_at:
                label: Date added
                type: datepicker
            details:
                label: Details
                type: text
                default: 'this is going to be saved in the db'

```

Would give values and so the developer can do the **_extra_** validation themselves.

~I think most developers would be ok with that?

**Comment by ayumihamsaki** [comment]
@LukeTowers 

I'm not sure of a good solution to that, it's a good question, I was thinking the same thing.

PHP Logic is this:

`[{"added_at":"","details":""}] = This array is NOT EMPTY, but contains BLANK VALUES.`

and

`[] = This array is EMPTY.`

This leaves us with the following options:

|  hide 1st array row | show 1st array row  |  show empty | show blank values |
|---|---|---|---|
| true | false | true | false |
| false | true | true | false |
| false | true | false | true |

Therefore, could we create a variable with two parameters, giving us an example code of say:

```
extra_information:
    type: repeater
    show: true, true
    form:
        fields:
           ..
```
 In the above code see the code line: `show: true, false` this would signal the following: show the 1st array row and set ALL Blank Values to Empty.

```
extra_information:
    type: repeater
    show: true, false
    form:
        fields:
           ..
```
 In the above code see the code line: `show: true, false` this would signal the following: show the 1st array row and LEAVE Blank Values in the array.

My idea, is based on the Laravel unique validator methos call, they do the same kind of thing there, see here: https://laravel.com/docs/5.6/validation#rule-unique

**Comment by ayumihamsaki** [comment]
~I'm thinking maybe it is not important to set the database field to `[]` at all. What do you think?

**Comment by LukeTowers** [comment]
@ayumihamsaki I'd agree with that. If the developer uses minItems then they can deal with that however they want, originally I was more concerned about default behaviour changing.

**Comment by ayumihamsaki** [comment]
~I changed my mind, I think it is not important to set the database field to `[]` at all. 

`php empty();`

The following values are considered to be empty:

- "" (an empty string)
- 0 (0 as an integer)
- 0.0 (0 as a float)
- "0" (0 as a string)
- NULL
- FALSE
- array() (an empty array)

Therefore it would not be necessary to check. I think best to leave out and just do your `minItems` only.

**Comment by LukeTowers** [comment]
@ayumihamsaki could you start work on a PR to implement minItems for the repeater?

**Comment by ayumihamsaki** [comment]
@LukeTowers I was thinking of paying a donation for this instead - if that's ok with you?

(I'm currently building a plugin which is as big as an app right now for everyone).

Off topic I have created a grid system for october backend and update the colour picker see screenshot:

![aaaaaaaaa](https://user-images.githubusercontent.com/17784082/43723599-ed025412-998f-11e8-8f16-4c177a1d3a0f.png)

Would these features be good to add to October? Just wondering to open another issue or not...?

**Comment by ayumihamsaki** [comment]
@LukeTowers I was thinking of paying a donation for this instead - if that's ok with you?

**Comment by LukeTowers** [comment]
@ayumihamsaki sure, I could implement this functionality for $300 USD.

**Comment by ayumihamsaki** [comment]
@LukeTowers I was thinking of donating $100 as this is a very minor enhacement to october.

**Comment by LukeTowers** [comment]
@ayumihamsaki when would you need it by?

**Comment by ayumihamsaki** [comment]
@LukeTowers I'm not bothered, whenever you got any free time. I just thought this issue was a bug in October CMS.

I get feedback from website owners about October CMS and they always tell me two main issues which are this issue and the media section being very slow.

So I thought I'd open this issue and I dont mind giving a donation of $100 to it because I use October 100% of my time and been using it for 3-4 years now.

So whenever you got some free time, let me now and I can pay you in paypal if you want via your email.

**Comment by LukeTowers** [comment]
@ayumihamsaki alright, that works for me :) $300 was for a priority change, but I can do it for the $100.

**Comment by LukeTowers** [comment]
@ayumihamsaki anything ending at luketowers.ca goes to the same inbox so it really doesn't matter 😄

**Comment by ayumihamsaki** [comment]
@LukeTowers I thought you were going to say anything (at) gmail.com goes to you lol. Send you an email, hope you get it thanks.

**Comment by LukeTowers** [comment]
@ayumihamsaki finished implementing `minItems` for the repeater, also added server side validation of minItems & maxItems.

You can send me the money by going to https://www.paypal.me/theluketowers/100usd

**Comment by ayumihamsaki** [comment]
@LukeTowers Many thanks have paid and sent you an email. Can I leave you to close this issue,.. Many thanks, look forward to seeing it in a future october version.

**Comment by LukeTowers** [comment]
@ayumihamsaki the issue will be automatically closed when the relevant commit is merged into master.

**Comment by danielbidala** [comment]
This feature does not seem to work in build 458!

My field config:

```
      addresses:
         type: repeater
         tab: Címek
         prompt: Új cím
         minItems: 1
         maxItems: 200
         form:
            fields:
               zip:
                  label: Irányítószám
                  type: text
               city:
                  label: Város
                  type: text
               address:
                  label: Cím
                  type: text
```

**Comment by LukeTowers** [comment]
@danielbidala open a new issue for that

**Comment by Arzon** [comment]
@danielbidala you can have a look at the following url 
https://octobercms.com/forum/post/repeater-default-value-backend-forms

**Comment by LukeTowers** [comment]
@bennothommo did the repeater styles addition change how minItems behaves?

**Comment by bennothommo** [comment]
@LukeTowers I don't believe it has.

**Comment by daftspunk** [comment]
This has been fixed in v2.1.8
